SV contraint {dist} not supported

TB Code ref:
https://github.com/mohamedtareq24/My_Verification_work/blob/main/FIFO/tb.sv





**Describe the bug/issue**
 

**To Reproduce**
Steps to reproduce the behavior:
1. python3 ../py_src/pyslint.py -t ../sv_tests/my_issue.sv

Add sample log below



Error:
```log
python3 ../py_src/pyslint.py -t ../../My_Verification_work-main/FIFO/fifo.sv >> fifo_spi_uart.log
python3 ../py_src/pyslint.py -t ../../My_Verification_work-main/FIFO/tb.sv   >> fifo_spi_uart.log
Traceback (most recent call last):
  File "/home/avs/vf/pyslint-main/exec_dir/../py_src/pyslint.py", line 313, in <module>
    cnst_arr_method_cast (scope_i)
  File "/home/avs/vf/pyslint-main/exec_dir/../py_src/pyslint.py", line 66, in cnst_arr_method_cast
    if (lv_cnst_i.expr.left.kind.name == 'InvocationExpression'):
AttributeError: 'pyslang.ExpressionOrDistSyntax' object has no attribute 'left'
make: *** [Makefile:67: fifo] Error 1

```

**Expected behavior**
No error 
 

**Additional context**

TB path:
https://github.com/mohamedtareq24/My_Verification_work/blob/main/SPI/tb.sv



Error:
python3 ../py_src/pyslint.py -t ../../My_Verification_work-main/SPI/spi.sv   >> fifo_spi_uart.log
python3 ../py_src/pyslint.py -t ../../My_Verification_work-main/SPI/tb.sv    >> fifo_spi_uart.log
Traceback (most recent call last):
  File "/home/avs/vf/pyslint-main/exec_dir/../py_src/pyslint.py", line 319, in <module>
    for m_i in (cu_scope.members):
AttributeError: 'pyslang.DataDeclarationSyntax' object has no attribute 'members'
make: *** [Makefile:69: fifo] Error 1

--- a/py_src/pyslint.py
+++ b/py_src/pyslint.py
@@ -67,9 +67,14 @@ def cnst_arr_method_cast (lv_cu_scope):
           # do not have left/right
           if (lv_cnst_i.expr.kind.name == 'UnaryLogicalNotExpression'):
             continue
+          # Fix for Issue 36, inside expressions
+          # do not have left/right
           if (lv_cnst_i.expr.kind.name == 'InsideExpression'):
             continue
-          print (lv_cnst_i.expr.kind.name)
+          # Fix for Issue 37, dist expressions
+          # do not have left/right
+          if (lv_cnst_i.expr.kind.name == 'ExpressionOrDist'):
+            continue
           if (lv_cnst_i.expr.left.kind.name == 'InvocationExpression'):
             lv_cnst_expr_s = lv_cnst_i.expr.left.__str__()
             lv_arr_red_methods = [".sum()", ".sum ()",
@@ -262,7 +267,8 @@ def pyslint_update_suffixes():
 
 def chk_name_style_prefix (lv_rule_id, lv_name, lv_exp_p):
   if (lv_name.startswith(lv_exp_p)):
-    print ('AF: Good naming: ', lv_name)
+    if (print_verbose):
+      print ('AF: Good naming: ', lv_name)
   else:
     msg = 'Improper naming of identifier: ' 
     msg += lv_name

--- a/sv_tests/test_sv_issue_37.sv
+++ b/sv_tests/test_sv_issue_37.sv
@@ -1,0 +1,514 @@
+class transaction;
+  
+  rand bit rd ,wr;
+  rand bit [7:0] data_in;
+  bit full, empty;
+  bit [7:0] data_out;
+  
+  constraint wr_rd {  
+    rd != wr;
+    wr dist {0 :/ 50 , 1:/ 50};
+    rd dist {0 :/ 50 , 1:/ 50};
+    
+  }
+ 
+  
+   constraint data_con {
+  data_in > 1; data_in < 5;  
+  }
+  
+  
+   function void display(input string tag);
+     $display("[%0s] : WR : %0b\t RD:%0b\t DATAWR : %0d\t DATARD : %0d\t FULL : %0b\t EMPTY : %0b @ %0t", tag, wr, rd, data_in, data_out, full, empty,$time);   
+  endfunction
+  
+  function transaction copy();
+    copy = new();
+    copy.rd = this.rd;
+    copy.wr = this.wr;
+    copy.data_in = this.data_in;
+    copy.data_out= this.data_out;
+    copy.full = this.full;
+    copy.empty = this.empty;
+  endfunction
+  
+endclass
+ 
+/*
+ 
+module tb;
+  
+  transaction tr;
+  
+  initial begin
+    tr = new();
+    tr.display("TOP");    
+  end
+  
+  
+endmodule
+ 
+*/
+ 
+ 
+ 
+ 
+ 
+class generator;
+  
+   transaction tr;
+   mailbox #(transaction) mbx;
+  
+   int count = 0;
+  
+   event next;  ///know when to send next transaction
+   event done;  ////conveys completion of requested no. of transaction
+   
+   
+  function new(mailbox #(transaction) mbx);
+      this.mbx = mbx;
+      tr=new();
+   endfunction; 
+  
+ 
+   task run(); 
+    
+     repeat(count)	 
+	     begin    
+           assert(tr.randomize) else $error("Randomization failed");	
+           mbx.put(tr.copy);
+           tr.display("GEN");
+           @(next);
+         end 
+     
+     
+     ->done;
+   endtask
+  
+  
+endclass
+ 
+/*
+  module tb;
+    
+    generator gen;
+    mailbox #(transaction) mbx;
+    
+    
+    
+    initial begin
+      mbx = new();
+      
+      gen = new(mbx);
+      
+      gen.count = 20;
+      gen.run();
+      
+      
+    end
+    
+ 
+ 
+ 
+ 
+ 
+    
+    
+    
+  endmodule
+  
+  */
+ 
+ 
+ 
+  
+  
+  
+class driver;
+  
+   virtual fifo_if fif;
+  
+   mailbox #(transaction) mbx;
+  
+   transaction datac;
+  
+   event next;
+  
+   
+ 
+    function new(mailbox #(transaction) mbx);
+      this.mbx = mbx;
+   endfunction; 
+  
+  ////reset DUT
+  task reset();
+    fif.rst <= 1'b1;
+    fif.rd <= 1'b0;
+    fif.wr <= 1'b0;
+    fif.data_in <= 0;
+    repeat(5) @(posedge fif.clock);
+    fif.rst <= 1'b0;
+    $display("[DRV] : DUT Reset Done");
+  endtask
+   
+  //////Applying RANDOM STIMULUS TO DUT
+  task run();
+    forever begin
+      mbx.get(datac);
+      
+      datac.display("DRV");
+      
+      fif.rd <= datac.rd;
+      fif.wr <= datac.wr;
+      fif.data_in <= datac.data_in;
+      repeat(2) @(posedge fif.clock);
+      ->next;
+    end
+  endtask
+  
+  
+endclass
+ 
+ 
+/*
+ 
+  module tb;
+    
+    generator gen;
+    
+    driver drv;
+    
+    event next;
+    
+    mailbox #(transaction) mbx;
+    
+    fifo_if fif();
+    
+    fifo dut (fif.clock, fif.rd, fif.wr,fif.full, fif.empty, fif.data_in, fif.data_out, fif.rst);
+    
+    initial begin
+      fif.clock <= 0;
+    end
+    
+    always #10 fif.clock <= ~fif.clock;
+    
+    initial begin
+      mbx = new();
+      
+      gen = new(mbx);
+      
+      gen.count = 20;
+      
+      drv = new(mbx);
+      
+      drv.fif = fif;
+      gen.next = next;
+      drv.next = next;
+      
+      
+    end
+    
+    initial begin
+      fork
+        gen.run();
+        drv.run();
+      join
+    end
+    
+    initial begin
+      #800;
+      $finish();      
+    end
+    
+    initial begin
+      $dumpfile("dump.vcd");
+      $dumpvars;
+    end
+    
+    
+  endmodule
+  
+ */
+ 
+ 
+ 
+ 
+class monitor;
+ 
+   virtual fifo_if fif;
+  
+   mailbox #(transaction) mbx;
+  
+   transaction tr;
+ 
+  
+  
+ 
+  
+    function new(mailbox #(transaction) mbx);
+      this.mbx = mbx;     
+   endfunction;
+  
+  
+  task run();
+    tr = new();
+    
+    forever begin
+      repeat(2) @(posedge fif.clock);
+      tr.wr = fif.wr;
+      tr.rd = fif.rd;
+      tr.data_in = fif.data_in;
+      tr.data_out = fif.data_out;
+      tr.full = fif.full;
+      tr.empty = fif.empty;
+      
+      
+      mbx.put(tr);
+      
+      tr.display("MON");
+ 
+    end
+    
+  endtask
+  
+ 
+  
+endclass
+ 
+/////////////////////////////////////////////////////
+ 
+ 
+class scoreboard;
+  
+   mailbox #(transaction) mbx;
+  
+   transaction tr;
+  
+   event next;
+  
+  bit [7:0] din[$];
+  bit[7:0] temp;
+  
+   function new(mailbox #(transaction) mbx);
+      this.mbx = mbx;     
+    endfunction;
+  
+  
+  task run();
+    
+  forever begin
+    
+    mbx.get(tr);
+    
+    tr.display("SCO");
+    
+    if(tr.wr == 1'b1)
+      begin 
+      din.push_front(tr.data_in);
+        $display("[SCO] : DATA STORED IN QUEUE :%0d", tr.data_in);
+      end
+    
+    if(tr.rd == 1'b1)
+      begin
+        if(tr.empty == 1'b0) begin 
+          
+          temp = din.pop_back();
+          
+          if(tr.data_out == temp)
+            $display("[SCO] : DATA MATCH");
+           else
+             $error("[SCO] : DATA MISMATCH");
+        end
+        else 
+          begin
+            $display("[SCO] : FIFO IS EMPTY");
+          end
+        
+        
+     end
+    
+    ->next;
+  end
+  endtask
+ 
+  
+endclass
+//////////////////////////////////////
+ 
+ 
+/*
+module tb;
+  
+  
+  monitor mon;
+  scoreboard sco;
+  
+  mailbox #(transaction) mbx;
+  
+  event next;
+  
+  
+   fifo_if fif();
+  
+   fifo dut (fif.clock, fif.rd, fif.wr,fif.full, fif.empty, fif.data_in, fif.data_out, fif.rst);
+    
+    initial begin
+      fif.clock <= 0;
+    end
+    
+    always #10 fif.clock <= ~fif.clock;
+ 
+  
+  
+  initial begin
+    mbx = new();
+    
+    mon = new(mbx);
+    sco = new(mbx);
+    
+    mon.fif = fif; 
+    
+    mon.next = next;
+    sco.next = next;
+  end
+  
+  initial begin
+    fork
+      mon.run();
+      sco.run();
+    join
+  end
+  
+      initial begin
+      #200;
+      $finish();      
+    end
+    
+    initial begin
+      $dumpfile("dump.vcd");
+      $dumpvars;
+    end
+  
+  
+  
+  
+  
+endmodule
+ 
+*/
+ 
+ 
+ 
+ 
+class environment;
+ 
+    generator gen;
+    driver drv;
+  
+    monitor mon;
+    scoreboard sco;
+  
+  mailbox #(transaction) gdmbx; ///generator + Driver
+    
+  mailbox #(transaction) msmbx; ///Monitor + Scoreboard
+ 
+  event nextgs;
+ 
+ 
+  virtual fifo_if fif;
+  
+  
+  function new(virtual fifo_if fif);
+ 
+    
+    
+    gdmbx = new();
+    gen = new(gdmbx);
+    drv = new(gdmbx);
+    
+    
+    
+    
+    msmbx = new();
+    mon = new(msmbx);
+    sco = new(msmbx);
+    
+    
+    this.fif = fif;
+    
+    drv.fif = this.fif;
+    mon.fif = this.fif;
+    
+    
+    gen.next = nextgs;
+    sco.next = nextgs;
+ 
+  endfunction
+  
+  
+  
+  task pre_test();
+    drv.reset();
+  endtask
+  
+  task test();
+  fork
+    gen.run();
+    drv.run();
+    mon.run();
+    sco.run();
+  join_any
+    
+  endtask
+  
+  task post_test();
+    wait(gen.done.triggered);  
+    $finish();
+  endtask
+  
+  task run();
+    pre_test();
+    test();
+    post_test();
+  endtask
+  
+  
+  
+endclass
+ 
+ 
+ 
+ 
+ 
+ 
+  module tb;
+    
+   
+    
+    fifo_if fif();
+    fifo dut (fif.clock, fif.rd, fif.wr,fif.full, fif.empty, fif.data_in, fif.data_out, fif.rst);
+    
+    initial begin
+      fif.clock <= 0;
+    end
+    
+    always #10 fif.clock <= ~fif.clock;
+    
+    environment env;
+    
+    
+    
+    initial begin
+      env = new(fif);
+      env.gen.count = 20;
+      env.run();
+    end
+      
+    
+    initial begin
+      $dumpfile("dump.vcd");
+      $dumpvars;
+    end
+   
+    
+  endmodule

--- a/sv_tests/test_sv_issue_37_1.sv
+++ b/sv_tests/test_sv_issue_37_1.sv
@@ -1,0 +1,41 @@
+
+ class distribution;
+   
+   // actually if we are not using ranges of values there is no diff between (:=) and (:/) operators
+   rand bit sel;// :=
+   rand bit en;// :/
+   
+   // but if we use ranges of values there we can recognize the diff
+   
+   rand bit [0:3] data1;//:=
+   rand bit [0:3] data2;//:/
+   
+   constraint one{
+     sel dist {0:=10, 1:=90};
+     en dist {0:/10, 1:/90};
+     data1 dist {0:=10, [1:3]:=60};
+     data2 dist {0:/10, [1:3]:/60};
+   }
+   
+ endclass
+   
+ module tb();
+   
+   distribution d;
+   
+   initial begin
+     d = new();
+     
+     for(int i = 0; i < 10; i++) begin
+       if(d.randomize()) begin 
+       end else begin
+         $display("randomization failed");
+       end
+     #10 
+       $display("sel = %0d, en = %0d ", d.sel, d.en);
+       $display("data1 = %0d, data2 = %0d ", d.data1, d.data2);
+       
+     end
+   end
+   
+ endmodule

